pkg/csi/service/wcp/controller.go

```diff
@@ -1299,8 +1302,138 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
+			}
+		}
+
+		// Check if snapshots number of this volume reaches the limit
+		snapshotList, _, err := common.QueryVolumeSnapshotsByVolumeID(ctx, c.manager.VolumeManager, volumeID,
+			common.QuerySnapshotLimit)
+		if err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to query snapshots of volume %s for the limit check. Error: %v", volumeID, err)
+		}
+
+		if len(snapshotList) >= maxSnapshotsPerBlockVolume {
+			return nil, logger.LogNewErrorCodef(log, codes.FailedPrecondition,
+				"the number of snapshots on the source volume %s reaches the configured maximum (%v)",
+				volumeID, c.manager.CnsConfig.Snapshot.GlobalMaxSnapshotsPerBlockVolume)
+		}
+
+		// the returned snapshotID below is a combination of CNS VolumeID and CNS SnapshotID concatenated by the "+"
+		// sign. That is, a string of "<UUID>+<UUID>". Because, all other CNS snapshot APIs still require both
+		// VolumeID and SnapshotID as the input, while corresponding snapshot APIs in upstream CSI require SnapshotID.
+		// So, we need to bridge the gap in vSphere CSI driver and return a combined SnapshotID to CSI Snapshotter.
+		snapshotID, snapshotCreateTimePtr, err := common.CreateSnapshotUtil(ctx, c.manager, volumeID, req.Name)
+		if err != nil {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create snapshot on volume %q: %v", volumeID, err)
+		}
+		snapshotCreateTimeInProto := timestamppb.New(*snapshotCreateTimePtr)
+
+		createSnapshotResponse := &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SizeBytes:      snapshotSizeInMB * common.MbInBytes,
+				SnapshotId:     snapshotID,
+				SourceVolumeId: volumeID,
+				CreationTime:   snapshotCreateTimeInProto,
+				ReadyToUse:     true,
+			},
+		}
+
+		log.Infof("CreateSnapshot succeeded for snapshot %s "+
+			"on volume %s size %d Time proto %+v Timestamp %+v Response: %+v",
+			snapshotID, volumeID, snapshotSizeInMB*common.MbInBytes, snapshotCreateTimeInProto,
+			*snapshotCreateTimePtr, createSnapshotResponse)
+
+		log.Infof("Attempting to annotate the associated volumesnapshot request with FCD and Snapshot " +
+			"annotation..")
+
+		volumeSnapshotName := req.Parameters[common.VolumeSnapshotNameKey]
+		volumeSnapshotNamespace := req.Parameters[common.VolumeSnapshotNamespaceKey]
+
+		log.Infof("Attempting to annotate volumesnapshot %s/%s with %s",
+			volumeSnapshotNamespace, volumeSnapshotName, snapshotID)
+		annotated, err := commonco.ContainerOrchestratorUtility.AnnotateVolumeSnapshot(ctx, volumeSnapshotName,
+			volumeSnapshotNamespace, map[string]string{common.VolumeSnapshotInfoKey: snapshotID})
+		if err != nil || !annotated {
+			log.Errorf("failed to annotate volumesnapshot %s, however, the snapshot %s was created"+
+				". Error: %v", volumeSnapshotName, snapshotID, err)
+		}
+		return createSnapshotResponse, nil
+	}
+
+	start := time.Now()
+	resp, err := createSnapshotInternal()
```

Comment:
Please follow the code convention for other controller APIs implementation that is updated recently. `faultType` is returned.

Reply:
This requires changes to our core CreateSnapshot and DeleteSnapshot calls in CnsManager, will address it later as it requires additional testing (since it affects even the vanilla case).